core/src/main/java/org/infinispan/interceptors/TxInterceptor.java

```diff
@@ -235,18 +249,73 @@ public Object visitSizeCommand(InvocationContext ctx, SizeCommand command) throw
    }
 
    @Override
-   public Object visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
-      return enlistReadAndInvokeNext(ctx, command);
-   }
-
-   @Override
-   public Object visitValuesCommand(InvocationContext ctx, ValuesCommand command) throws Throwable {
-      return enlistReadAndInvokeNext(ctx, command);
+   public CacheSet<K> visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
+      CacheSet<K> set = (CacheSet<K>) enlistReadAndInvokeNext(ctx, command);
+      if (ctx.isInTxScope()) {
+         return new AbstractDelegatingKeyCacheSet(getCacheWithFlags(cache, command), set) {
+            @Override
+            public CloseableIterator<K> iterator() {
+               return new TransactionAwareKeyCloseableIterator<>(super.iterator(),
+                       (TxInvocationContext<LocalTransaction>) ctx, cache);
+            }
+
+            @Override
+            public CloseableSpliterator<K> spliterator() {
+               Spliterator<K> parentSpliterator = super.spliterator();
+               long estimateSize = parentSpliterator.estimateSize() + ctx.getLookedUpEntries().size();
+               // This is an overestimate for size if we have looked up entries that don't map to this node
+               return new IteratorAsSpliterator.Builder<>(iterator())
+                       .setEstimateRemaining(estimateSize < 0L ? Long.MAX_VALUE : estimateSize)
+                       .setCharacteristics(Spliterator.CONCURRENT | Spliterator.DISTINCT | Spliterator.NONNULL)
+                       .get();
+            }
+
+            @Override
+            public int size() {
+               long size = stream().count();
+               if (size > Integer.MAX_VALUE) {
+                  return Integer.MAX_VALUE;
+               }
+               return (int) size;
+            }
+         };
+      }
+      return set;
    }
 
    @Override
-   public Object visitEntrySetCommand(InvocationContext ctx, EntrySetCommand command) throws Throwable {
-      return enlistReadAndInvokeNext(ctx, command);
+   public CacheSet<CacheEntry<K, V>> visitEntrySetCommand(InvocationContext ctx, EntrySetCommand command) throws Throwable {
+      CacheSet<CacheEntry<K, V>> set = (CacheSet<CacheEntry<K, V>>) enlistReadAndInvokeNext(ctx, command);
+      if (ctx.isInTxScope()) {
+         return new AbstractDelegatingEntryCacheSet<K, V>(getCacheWithFlags(cache, command), set) {
```

Comment:
It is identical except for the class it is extending.  And unfortunately they require calling super methods so I can't remove them, however I could remove a bit of code.